src/cache.rs

```diff
@@ -270,33 +422,40 @@ impl<T: Hash + PartialEq> Constraint<T> {
     where
         F: FnMut(&T) -> u128,
     {
-        let calls = self.0.borrow();
-        ACCELERATOR.with(|accelerator| {
-            let mut map = accelerator.borrow_mut();
-            calls.iter().all(|entry| {
-                *map.entry((id, entry.both)).or_insert_with(|| f(&entry.args))
-                    == entry.ret
+        let accelerator = accelerator(id);
+        let inner = self.0.read();
+        if let Some(accelerator) = accelerator {
+            let mut map = accelerator.lock();
+            inner.values().all(|entry| {
+                *map.entry(entry.both).or_insert_with(|| f(&entry.args)) == entry.ret
```

Comment:
I _think_ `entry.args_hash` instead of `entry.both` would actually suffice here and then `both` could be removed. I am not 100% sure though.

Reply:
It does seem to work indeed. I wasn't too sure hence why I did it this way.